ord_schema/dataset_templating_test.py

```diff
@@ -0,0 +1,108 @@
+
+        message = reaction_pb2.Reaction()
+        dummy_input = message.inputs['in']
+        outcome = message.outcomes.add()
+        component = dummy_input.components.add()
+        component.identifiers.add(type='SMILES', value='CCO')
+        component.is_limiting = True
+        component.mass.value = 1
+        component.mass.units = reaction_pb2.Mass.GRAM
+        outcome.conversion.value = 75
+        self.valid_reaction = message
+
+    def test_valid_templating(self):
+        template_string = text_format.MessageToString(self.valid_reaction)
+        template_string = template_string.replace('value: "CCO"',
+                                                  'value: "$my_smiles$"')
+        template_string = template_string.replace('value: 75',
+                                                  'value: $conversion$')
+        df = pd.DataFrame.from_dict({
+            '$my_smiles$': ['CCO', 'CCCO', 'CCCCO'],
+            '$conversion$': ['75', '50', '30'],
+        })
+        dataset = dataset_templating.generate_dataset(template_string, df)
+        exptected_reactions = []
+        for smiles, conversion in zip(['CCO', 'CCCO', 'CCCCO'], [75, 50, 30]):
+            reaction = reaction_pb2.Reaction()
+            reaction.CopyFrom(self.valid_reaction)
+            reaction.inputs['in'].components[0].identifiers[0].value = smiles
+            reaction.outcomes[0].conversion.value = conversion
+            exptected_reactions.append(reaction)
+        expected_dataset = dataset_pb2.Dataset(reactions=exptected_reactions)
+        self.assertEqual(dataset, expected_dataset)
+
+        df = pd.DataFrame.from_dict({
+            'my_smiles': ['CCO', 'CCCO', 'CCCCO'],
+            'conversion': ['75', '50', '30'],
+        })
+        dataset = dataset_templating.generate_dataset(template_string, df)
+        self.assertEqual(dataset, expected_dataset)
+
+    def test_invalid_templating(self):
+        template_string = text_format.MessageToString(self.valid_reaction)
+        template_string = template_string.replace('value: "CCO"',
+                                                  'value: "$my_smiles$"')
+        template_string = template_string.replace('value: 75',
+                                                  'value: $conversion$')
+        df = pd.DataFrame.from_dict({
+            '$my_smiles$': ['CCO', 'CCCO', 'CCCCO'],
+            '$conversion$': ['75', '50', '-5'],
+        })
+        exptected_reactions = []
+        for smiles, conversion in zip(['CCO', 'CCCO', 'CCCCO'], [75, 50, -5]):
+            reaction = reaction_pb2.Reaction()
+            reaction.CopyFrom(self.valid_reaction)
+            reaction.inputs['in'].components[0].identifiers[0].value = smiles
+            reaction.outcomes[0].conversion.value = conversion
+            exptected_reactions.append(reaction)
+        expected_dataset = dataset_pb2.Dataset(reactions=exptected_reactions)
+        with self.assertRaisesRegex(ValueError, 'is not valid'):
```

Comment:
can you make this error regex a bit more specific so we know where it's failing?